Liblitedram cleanup

michalsieron opened the pull request:
This PR introduces a number of changes that clean up existing liblitedram code.
It extracts existing functions like `write_rst_dqs_delay` or `write_inc_dq_delay` to separate file - `accessors.h`.
As all those functions are used in the same way:
- select module (and optionally specific line)
- increment or reset current delay
- deselect module and line

a new function was added `sdram_leveling_action`, which abstracts this away.

Support for x4 IC configuration was fixed and ability to train individual DQ lines was added.

Additionally, comments were added to `#endif`s, which indicate conditions of their `#if`s.

All changes were tested on Digilent Arty A7 board and [Antmicro Data Center DRAM Tester](https://github.com/litex-hub/litex-boards/blob/master/litex_boards/targets/antmicro_datacenter_ddr4_test_board.py)

enjoy-digital commented on litex/soc/software/liblitedram/sdram.c:
```diff
@@ -344,90 +361,111 @@ static unsigned int sdram_write_read_check_test_pattern(int module, unsigned int
-#endif
-	for(i=0;i<SDRAM_PHY_DELAY_JUMP;i++) {
-		delay += 1;
-		inc_delay(module);
```
Can you explain the reason to remove the delay jump?

enjoy-digital commented:
Thanks @michalsieron, I'll plan some time to closely review this and combine it with other improvements of https://github.com/enjoy-digital/litex/pull/1499. For now can you or @mtdudek explain the reason to remove the delay jump?

michalsieron commented:
I rebased it onto the current master branch, however I still have to check the issues with the Ultrascale.

michalsieron commented:
@enjoy-digital I identified and fixed the issue with the Ultrascale FPGAs.

In [write_rst_dq_delay](https://github.com/enjoy-digital/litex/commit/9de97c740e671196a52ba6e57b3ba1e0a1995991#diff-c4cb7cce8801994caeca9e6ade04313c730bdb0320e070fcb7498f930608f03aR41-R53) accessor,
```
	while (ddrphy_wdly_dqs_inc_count_read() != 0) {
		ddrphy_wdly_dq_inc_write(1);
		cdelay(100);
	}
```
which replaced [sdram_write_dq_dqs_training_rst_delay](https://github.com/enjoy-digital/litex/commit/9de97c740e671196a52ba6e57b3ba1e0a1995991#diff-dfaac55cd2995171fc9020483161a662fc8f941dfa7dbb93aa998ad4c4447fc5L1277-L1297), there was a mistake in the `while` loop condition.
```
	int dq_count = ddrphy_wdly_dqs_inc_count_read();
	while (dq_count != SDRAM_PHY_DELAYS) {
		ddrphy_wdly_dq_inc_write(1);
		cdelay(100);
		dq_count++;
	}
```

During the cleanup, `ddrphy_wdly_dqs_inc_count_read` was treated as it was reading DQ delay count and variable `dq_count` have been removed.

This resulted in an infinite loop, which caused the BIOS to stuck.

---

To find this, I had to split some commits. This is a good thing, as they should be more digestible now.
I also moved the accessors from a header file to a proper `.c` file.

enjoy-digital commented:
Great, thanks @michalsieron!
